### PR TITLE
Implement globstar support for Cassandra driver

### PR DESCRIPTION
**Pull request by dpanth3r**
Solves #96

---
This change is [Reviewable](https://reviewable.io/reviews/criteo/biggraphite/149)

### CODE REVIEW

File: biggraphite/drivers/cassandra.py [hunk 1]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
...
+
+        if len(metric_names) > self.max_metrics_per_glob:
+            msg = "%s yields more than %d results" % (glob, self.max_metrics_per_glob)
+            raise TooManyMetrics(msg)
+
+        metric_names.sort()
+        return metric_names
+
+    def __build_select_metric_names_query(self, table, components, glob, is_raw=False):
         where_parts = [
```
Comment (rudy-6-4) [inline biggraphite/drivers/cassandra.py]: can you doc ?
does it return all metrics in that path ?

File: biggraphite/drivers/cassandra.py [hunk 2]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
             msg = "Metric globs can have a maximum of %d dots" % _COMPONENTS_MAX_LEN - 2
             raise bg_accessor.InvalidGlobError(msg)
 
+        queries = []
+
+        # Handle globstars (wildcard on any number of components)
+        if "**" in components:
```
Comment (rudy-6-4) [inline biggraphite/drivers/cassandra.py]: can you share the MULTI_WILDCARD = '**' once for all ?

File: biggraphite/drivers/cassandra.py [hunk 3]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
...
+            # otherwise we have to use iterative deepening with wildcards.
+            gs_index = components.index("**")
+            if gs_index == len(components) - 2:
+                gs_query = self.__build_select_metric_names_query(
+                    table, components[:-2], glob, is_raw=True)
+                queries.append(gs_query)
+            else:
+                prefix = components[0:gs_index]
+                suffix = components[gs_index+1:]
+                max_wildcards = min(
```
Comment (rudy-6-4) [inline biggraphite/drivers/cassandra.py]: max_wildcard_for_globstar is a better name for max_wildcards

File: biggraphite/drivers/cassandra.py [hunk 4]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
...
+        else:
+            query = self.__build_select_metric_names_query(table, components, glob)
+            queries.append(query)
+
+        metric_names = []
+        try:
+            for query in queries:
+                for row in self._execute(query):
+                    metric_names.append(row[0])
+                if len(metric_names) > self.max_metrics_per_glob:
```
Comment (rudy-6-4) [inline biggraphite/drivers/cassandra.py]: why break to raise later ?

File: biggraphite/drivers/cassandra.py [hunk 5]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
...
+                    break
+        except Exception as e:
```
Comment (rudy-6-4) [inline biggraphite/drivers/cassandra.py]: can you write the exception types?

File: biggraphite/drivers/cassandra.py [hunk 6]
```diff
@@ -749,32 +760,81 @@ def __glob_names(self, table, glob):
...
+                    gs_query = self.__build_select_metric_names_query(table, gs_components, glob)
+                    queries.append(gs_query)
+        else:
+            query = self.__build_select_metric_names_query(table, components, glob)
+            queries.append(query)
+
+        metric_names = []
+        try:
+            for query in queries:
+                for row in self._execute(query):
```
Comment (dpanth3r) [inline biggraphite/drivers/cassandra.py]: agreed, but I feel this would diminish readability

**Pull request merged**